lua/nvim-treesitter/query_predicates.lua

```diff
@@ -64,5 +74,16 @@ query.add_predicate('is?', function(match, pattern, bufnr, pred)
   return vim.tbl_contains(types, kind)
 end)
 
+query.add_predicate('has-type?', function(match, pattern, bufnr, pred)
+  if not valid_args(pred[1], pred, 2) then return end
+
+  local node = match[pred[2]]
+  local types = {unpack(pred, 3)}
+
+  if not node then return true end
+
+  return vim.tbl_contains(types, node:type())
+end)
+
```

Comment:
No no no, use `[]` syntax for that.

Reply:
what do you mean?

Reply:
@vigoux I didn't not quite understand what your opinion was to solve this. It's about having the same children tree but with different types as parents vs `[ ]` where you can have alternate subtrees but need to specify each subtree completely.

Reply:
Maybe we can have that as a workaround for some time, but this should be raised as an issue upstream. For me, the whole point of queries is to avoid strange predicates matching the type of a node.

Reply:
Usually, the purpose of of using Lisp-syntax is to have no syntax at all. I find it much more elegant to define a predicate for the current application on hand (only local to this plugin) than to pollute upstream with dozens of special cases.

`has-type` (not used in this PR) could be expressed by `[ ]` https://github.com/tree-sitter/tree-sitter/issues/759#issuecomment-706551332 There is however some problems with how `[ ]` was designed which make it difficult where to place the common children. We can easily transition to the official solution.